src/main/java/com/timgroup/statsd/StatsDProcessor.java

```diff
@@ -49,7 +48,15 @@
                 .onMalformedInput(CodingErrorAction.REPLACE)
                 .onUnmappableCharacter(CodingErrorAction.REPLACE);
 
-        public abstract void run();
+        public final void run() {
+            try {
+                processLoop();
+            } finally {
+                endSignal.countDown();
+            }
+        }
+
+        protected abstract void processLoop();
```

Comment:
Given that the worker is being interrupted to halt the worker, should we not invert control so the loop handles the interrupts? What if the implementation of this method catches the interrupt and resets the interrupted status?

Reply:
Yes, I'd like to refactor the worker loop further once this change is in (I'd prefer to keep that a separate PR though)